CRM/Mailing/BAO/Mailing.php

```diff
@@ -1140,9 +1140,12 @@ public function compose(
 
     // push the tracking url on to the html email if necessary
     if ($this->open_tracking && $html) {
-      array_push($html, "\n" . '<img src="' . CRM_Utils_System::externUrl('extern/open', "q=$event_queue_id")
-        . '" width="1" height="1" alt="" border="0">'
+      $tracking_url = $config->userSystem->languageNegotiationURL(
+        $config->userFrameworkBaseURL . Civi::settings()->get('userFrameworkResourceURL') . "/extern/open.php?q=$event_queue_id",
+        FALSE,
+        TRUE
```

Comment:
Aaah, so I now see that that #15336 and https://github.com/civicrm/org.civicrm.flexmailer/pull/47 were close cousins. Some of the comments about the settings/paths from there would also apply here.

Assuming that we want to get the URLs for the standalone scripts working... the fix should be intrinsic to  `CRM_Utils_System::externUrl()` so that any consumer of that function gets a correct URL. That might mean patching `externUrl()` or adding `hook_civicrm_alterExternUrl()` listener. (I can't really say precisely - because I don't fully grok the configuration it's running in. I'd need [more example data](https://lab.civicrm.org/dev/mail/issues/17#note_36907) to relate better.)

> This is a workaround until those endpoints for tracking links will be converted to something else using proper CMS routes, see dev/cloud-native#16

So... do you still want to get the URLs for standalone scripts working in this configuration... or are we making enough progress on [migrating to the normal router](https://github.com/civicrm/civicrm-core/pull/17312)?

Reply:
I'm not sure I'm the right one to answer your last question. We needed something that worked and I came up with what I suggested in this PR, which obviously is not 100% correct either, as per your comments in https://github.com/civicrm/org.civicrm.flexmailer/pull/47#discussion_r424992520. I don't have enough insight to decide what would be the better approach.